packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1048 @@
+ * @this Blockly.Block
+ */
+const controlsIfHelper = function() {
+  this.topInput_ = this.getInput('IF0');
+  this.topInput_.insertFieldAt(0, new FieldPlus(), 'PLUS');
+};
+
+Blockly.Extensions.unregister('controls_if_mutator');
+Blockly.Extensions.registerMutator('controls_if_mutator',
+    controlsIfMutator, controlsIfHelper);
+
+const textJoinMutator = {
+  /**
+   * Number of text inputs on this block.
+   * @type {number}
+   */
+  itemCount_: 0,
+
+  /**
+   * Creates XML to represent number of inputs.
+   * @return {!Element} XML storage element.
+   * @this Blockly.Block
+   */
+  mutationToDom: function() {
+    const container = Blockly.utils.xml.createElement('mutation');
+    container.setAttribute('items', this.itemCount_);
+    return container;
+  },
+  /**
+   * Parses XML to restore the inputs.
+   * @param {!Element} xmlElement XML storage element.
+   * @this Blockly.Block
+   */
+  domToMutation: function(xmlElement) {
+    const targetCount = parseInt(xmlElement.getAttribute('items'), 10);
+    this.updateShape_(targetCount);
+  },
+
+  /**
+   * Adds inputs to the block until the block reaches the target input count.
+   * @param {number} targetCount The number of inputs the block should have.
+   * @this Blockly.Block
+   * @private
+   */
+  updateShape_: function(targetCount) {
+    while (this.itemCount_ < targetCount) {
+      this.addPart_();
+    }
+    while (this.itemCount_ > targetCount) {
+      this.removePart_();
+    }
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the plus image. Adds an input to the block and updates the
+   * state of the minus.
+   * @this Blockly.Block
+   */
+  plus: function() {
+    this.addPart_();
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the minus image. Removes the input at the end of the block and
+   * updates the state of the minus.
+   * @this Blockly.Block
+   */
+  minus: function() {
+    if (this.itemCount_ == 0) {
+      return;
+    }
+    this.removePart_();
+    this.updateMinus_();
+  },
+
+  // To properly keep track of indices we have to increment before/after adding
+  // the inputs, and decrement the opposite.
+  // Because we want our first item to be ADD0 (not ADD1) we increment after.
+
+  /**
+   * Adds an input to the end of the block. If the block currently has no
+   * inputs it updates the top 'EMPTY' input to receive a block.
+   * @this Blockly.Block
+   * @private
+   */
+  addPart_: function() {
+    if (this.itemCount_ == 0) {
+      if (this.getInput('EMPTY')) {
+        // TODO: I don't think this should throw errors. It would be nice if
+        //   It returned a boolean instead.
+        this.removeInput('EMPTY');
+      }
+      this.topInput_ = this.appendValueInput('ADD' + this.itemCount_)
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(Blockly.Msg['TEXT_JOIN_TITLE_CREATEWITH']);
+    } else {
+      this.appendValueInput('ADD' + this.itemCount_);
+    }
+    this.itemCount_++;
```

Comment:
Put a comment here that says "// Input names are zero-indexed (ADD0, ADD1, etc)" and another before line 389 that says the same.

And remove the comment from 356-358

Reply:
Similar story for the related comments for the if block--delete the comment hanging out between functions (lines 206-209) in favor of comments on the lines that need to be explained.

Reply:
How does "Because item inputs are 0-index we decrement first, increment last" sound to you?

I want to get across the idea that we need to do one (decrementing or incrementing) at one end, and the other at the other end.

This is because when I was writing this I would drop it for a few weeks and come back later. I kept having this problem where I would be like "why didn't I make them match!?" so I would move them and break everything. I'd like to keep a reminder that they need to be done opposite eachother in there somewhere.